PWGCF/TwoParticleCorrelations/Tasks/lambdaR2Correlation.cxx

```diff
@@ -459,69 +469,123 @@ struct lambdaCorrelationAnalysis {
-
-    float rap2 = p2.y();
-    float phi2 = p2.phi();
+    if (cfg_do_eta_analysis) {
+      rapbin1 = static_cast<int>((p1.eta() - kminrap) / rapbinwidth);
+      rapbin2 = static_cast<int>((p2.eta() - kminrap) / rapbinwidth);
+    } else {
+      rapbin1 = static_cast<int>((p1.y() - kminrap) / rapbinwidth);
+      rapbin2 = static_cast<int>((p2.y() - kminrap) / rapbinwidth);
+    }
 
-    int rapbin1 = static_cast<int>((rap1 - kminrap) / rapbinwidth);
-    int phibin1 = static_cast<int>(phi1 / phibinwidth);
-    int rapbin2 = static_cast<int>((rap2 - kminrap) / rapbinwidth);
-    int phibin2 = static_cast<int>(phi2 / phibinwidth);
+    int phibin1 = static_cast<int>(p1.phi() / phibinwidth);
+    int phibin2 = static_cast<int>(p2.phi() / phibinwidth);
 
     if (rapbin1 >= 0 && rapbin2 >= 0 && phibin1 >= 0 && phibin2 >= 0 && rapbin1 < nrapbins && rapbin2 < nrapbins && phibin1 < nphibins && phibin2 < nphibins) {
 
       int rapphix = rapbin1 * nphibins + phibin1;
       int rapphiy = rapbin2 * nphibins + phibin2;
 
-      histos.fill(HIST(sub_dir_type[mode]) + HIST(sub_dir_hist[hist]), rapphix + 0.5, rapphiy + 0.5);
+      histos.fill(HIST(sub_dir_type[mass_win]) + HIST(sub_dir_hist[part_pair]), rapphix + 0.5, rapphiy + 0.5);
+    }
+  }
+
+  template <ParticleType part, MassWindowType mass_window, typename T>
+  void analyseSingles(T const& track, std::vector<std::vector<float>> v_mass_win)
+  {
+
+    // apply rapidity cut
+    if (cfg_do_eta_analysis) {
+      if (std::abs(track.eta()) > cfg_Rap_Max) {
+        return;
+      }
+    } else {
+      if (std::abs(track.y()) > cfg_Rap_Max) {
+        return;
+      }
+    }
+
+    // apply mass window acceptance
+    if (track.mass() <= v_mass_win[mass_window][kMin] || track.mass() >= v_mass_win[mass_window][kMax]) {
+      return;
+    }
+
+    fillHistos<part>(track);
+    partSingle<part, mass_window>(track);
+  }
+
+  template <ParticlePair part_pair, MassWindowType mass_window, bool same, typename T>
+  void analysePairs(T const& trk_1, T const& trk_2, std::vector<std::vector<float>> v_mass_win)
+  {
+
+    // apply rapidity cut
+    if (cfg_do_eta_analysis) {
+      if (std::abs(trk_1.eta()) > cfg_Rap_Max || std::abs(trk_2.eta()) > cfg_Rap_Max) {
+        return;
+      }
+    } else {
+      if (std::abs(trk_1.y()) > cfg_Rap_Max || std::abs(trk_2.y()) > cfg_Rap_Max) {
+        return;
+      }
+    }
+
+    // apply mass window acceptance
+    if (trk_1.mass() <= v_mass_win[mass_window][kMin] || trk_1.mass() >= v_mass_win[mass_window][kMax]) {
+      return;
+    }
+
+    if (trk_2.mass() <= v_mass_win[mass_window][kMin] || trk_2.mass() >= v_mass_win[mass_window][kMax]) {
+      return;
+    }
+
+    if constexpr (!same) {
+      partPair<part_pair, mass_window>(trk_1, trk_2);
+    } else {
+      if (trk_1.index() != trk_2.index() && trk_1.postrackid() != trk_2.postrackid() && trk_1.negtrackid() != trk_2.negtrackid()) {
+        partPair<part_pair, mass_window>(trk_1, trk_2);
```

Comment:
This is the most probable case when analyzing same particles and it will require the execution of three conditionals on a per pair level
It will be good to revisit this sequence of conditionals if there are problems with the required execution time